integration_tests/src/validation/audio.rs

```diff
@@ -34,75 +124,167 @@ pub fn validate(
 
-        let diff_pitch_left = f64::abs(expected_pitch_left - actual_pitch_left);
-        let diff_pitch_right = f64::abs(expected_pitch_right - actual_pitch_right);
+        let (expected_result_left, expected_result_right, actual_result_left, actual_result_right) =
+            analyze_samples(actual_samples, expected_samples, sample_rate)?;
 
-        if diff_pitch_left > allowed_error as f64 || diff_pitch_right > allowed_error as f64 {
-            let pts_start = time_range.start.as_micros();
-            let pts_end = time_range.end.as_micros();
+        let left_result = AnalyzeResult::compare(
+            &actual_result_left,
+            &expected_result_left,
+            &tolerance,
+            interval.first_sample,
+            Channel::Left,
+        );
+        let right_result = AnalyzeResult::compare(
+            &actual_result_right,
+            &expected_result_right,
+            &tolerance,
+            interval.first_sample,
+            Channel::Right,
+        );
 
-            return Err(anyhow::anyhow!(
-                "Audio mismatch. Time range: ({pts_start}, {pts_end}), Expected: ({}, {}) Actual: ({}, {})",
-                expected_pitch_left, expected_pitch_right,
-                actual_pitch_left, actual_pitch_right
-            ));
+        if let Err(err) = left_result {
+            error!("{err}");
+            failed_batches += 1;
+        }
+        if let Err(err) = right_result {
+            error!("{err}");
+            failed_batches += 1;
         }
     }
-    Ok(())
+
+    if failed_batches <= allowed_failed_batches {
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!("Test failed!"))
+    }
 }
 
-fn find_sample_batches(
-    samples: &[AudioSampleBatch],
-    time_range: Range<Duration>,
-) -> Vec<AudioSampleBatch> {
-    samples
-        .iter()
-        .filter(|s| time_range.contains(&s.pts))
-        .cloned()
-        .collect()
+fn find_samples(samples: &[f32], interval: SamplingInterval) -> Vec<f32> {
+    let first_sample = interval.first_sample;
+    let last_sample = interval.first_sample + interval.samples;
+    if first_sample >= samples.len() {
+        vec![0.0; interval.samples]
+    } else if last_sample > samples.len() {
+        let mut batch = samples[first_sample..].to_vec();
+        batch.resize(interval.samples, 0.0);
+        batch
+    } else {
+        samples[first_sample..last_sample].to_vec()
+    }
 }
 
-fn pitch_from_sample_batch(
-    sample_batch: Vec<AudioSampleBatch>,
+fn analyze_samples(
+    actual_samples: Vec<f32>,
+    expected_samples: Vec<f32>,
     sample_rate: u32,
-) -> Result<(f64, f64)> {
-    fn get_pitch(samples: &[f64], sample_rate: u32) -> Result<f64> {
-        if samples.is_empty() {
-            return Err(anyhow::anyhow!("No samples"));
+) -> Result<(AnalyzeResult, AnalyzeResult, AnalyzeResult, AnalyzeResult)> {
+    fn split_samples(samples: Vec<f32>) -> (Vec<f32>, Vec<f32>) {
```

Comment:
let's leave it as it is, but in the future, try to avoid this pattern, even though you are extracting logic to functions the fact that it is defined inside it still makes `analyze_samples` very complex.

If you want to extract those functions then just write them at top level,
If you wanted them to be private, because they are coupled stronglly witn analyze_samples you can  move everything to `validation/audio/analyze_samples.rs`
Even in this approach if there some of those functions can be described without the context of analyze druation, then maybe just those could be on top level.